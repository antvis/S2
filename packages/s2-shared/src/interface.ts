import type {
  S2DataConfig,
  S2Options,
  CellScrollPosition,
  TargetCellInfo,
  ResizeParams,
  Node,
  SpreadSheet,
  ThemeCfg,
  ViewMeta,
  LayoutResult,
  SortParams,
  DataCell,
  Data,
  GEvent,
  HiddenColumnsInfo,
  CollapsedRowsType,
  DataType,
  ResizeInfo,
  S2CellType,
  TooltipOperatorOptions,
  S2RenderOptions,
  S2MountContainer,
  CellMeta,
  TooltipContentType,
  Pagination,
} from '@antv/s2';

// 是否开启自适应宽高，并指定容器
export type Adaptive =
  | boolean
  | {
      width?: boolean;
      height?: boolean;
      getContainer?: () => HTMLElement;
    };

export type SheetType =
  | 'pivot'
  | 'table'
  | 'gridAnalysis'
  | 'strategy'
  | 'editable';

/** render callback */
export type SheetUpdateCallback = (params: S2RenderOptions) => S2RenderOptions;

export interface BaseSheetComponentProps<
  PartialDrillDown = unknown,
  Header = unknown,
  Options = S2Options<TooltipContentType, Pagination>,
> {
  sheetType?: SheetType;
  spreadsheet?: (
    container: S2MountContainer,
    dataCfg: S2DataConfig,
    options: Options,
  ) => SpreadSheet;
  dataCfg: S2DataConfig;
  options?: Options;
  loading?: boolean;
  partDrillDown?: PartialDrillDown;
  adaptive?: Adaptive;
  showPagination?:
    | boolean
    | {
        onShowSizeChange?: (current: number, pageSize: number) => void;
        onChange?: (current: number, pageSize: number) => void;
      };
  themeCfg?: ThemeCfg;
  header?: Header;
  /** @deprecated 1.29.0 已废弃, 请使用 onMounted 代替 */
  getSpreadSheet?: (spreadsheet: SpreadSheet) => void;
  /** 底表 render callback */
  onSheetUpdate?: SheetUpdateCallback;

  // ============== Row Cell ====================
  onRowCellHover?: (data: TargetCellInfo) => void;
  onRowCellClick?: (data: TargetCellInfo) => void;
  onRowCellDoubleClick?: (data: TargetCellInfo) => void;
  onRowCellContextMenu?: (data: TargetCellInfo) => void;
  onRowCellMouseDown?: (data: TargetCellInfo) => void;
  onRowCellMouseUp?: (data: TargetCellInfo) => void;
  onRowCellMouseMove?: (data: TargetCellInfo) => void;
  onRowCellCollapseTreeRows?: (params: {
    id: number;
    isCollapsed: boolean;
    node: Node;
  }) => void;
  onRowCellScroll?: (position: CellScrollPosition) => void;

  // ============== Col Cell ====================
  onColCellHover?: (data: TargetCellInfo) => void;
  onColCellClick?: (data: TargetCellInfo) => void;
  onColCellDoubleClick?: (data: TargetCellInfo) => void;
  onColCellContextMenu?: (data: TargetCellInfo) => void;
  onColCellMouseDown?: (data: TargetCellInfo) => void;
  onColCellMouseUp?: (data: TargetCellInfo) => void;
  onColCellMouseMove?: (data: TargetCellInfo) => void;

  // ============== Data Cell ====================
  onDataCellHover?: (data: TargetCellInfo) => void;
  onDataCellClick?: (data: TargetCellInfo) => void;
  onDataCellDoubleClick?: (data: TargetCellInfo) => void;
  onDataCellContextMenu?: (data: TargetCellInfo) => void;
  onDataCellMouseDown?: (data: TargetCellInfo) => void;
  onDataCellMouseUp?: (data: TargetCellInfo) => void;
  onDataCellMouseMove?: (data: TargetCellInfo) => void;
  onDataCellTrendIconClick?: (meta: ViewMeta) => void;
  onDataCellBrushSelection?: (brushRangeDataCells: DataCell[]) => void;
  onDataCellSelectMove?: (metas: CellMeta[]) => void;

  // ============== Corner Cell ====================
  onCornerCellHover?: (data: TargetCellInfo) => void;
  onCornerCellClick?: (data: TargetCellInfo) => void;
  onCornerCellDoubleClick?: (data: TargetCellInfo) => void;
  onCornerCellContextMenu?: (data: TargetCellInfo) => void;
  onCornerCellMouseDown?: (data: TargetCellInfo) => void;
  onCornerCellMouseUp?: (data: TargetCellInfo) => void;
  onCornerCellMouseMove?: (data: TargetCellInfo) => void;

  // ============== Merged Cells ====================
  onMergedCellsHover?: (data: TargetCellInfo) => void;
  onMergedCellsClick?: (data: TargetCellInfo) => void;
  onMergedCellsDoubleClick?: (data: TargetCellInfo) => void;
  onMergedCellsContextMenu?: (data: TargetCellInfo) => void;
  onMergedCellsMouseDown?: (data: TargetCellInfo) => void;
  onMergedCellsMouseUp?: (data: TargetCellInfo) => void;
  onMergedCellsMouseMove?: (data: TargetCellInfo) => void;

  // ============== Sort ====================
  onRangeSort?: (params: SortParams) => void;
  onRangeSorted?: (event: GEvent) => void;

  // ============== Filter ====================
  onRangeFilter?: (data: {
    filterKey: string;
    filteredValues: string[];
  }) => void;
  onRangeFiltered?: (data: DataType[]) => void;

  // ============== Layout ====================
  onLayoutAfterHeaderLayout?: (layoutResult: LayoutResult) => void;
  onLayoutPagination?: (data: {
    pageSize: number;
    pageCount: number;
    total: number;
    current: number;
  }) => void;
  /** @deprecated 已废弃, 请使用 S2Event.GLOBAL_SCROLL 代替 */
  onLayoutCellScroll?: (position: CellScrollPosition) => void;
  onLayoutCollapseRows?: (data: CollapsedRowsType) => void;
  onLayoutAfterCollapseRows?: (data: CollapsedRowsType) => void;
  onCollapseRowsAll?: (hierarchyCollapse: boolean) => void;
  onLayoutColsExpanded?: (node: Node) => void;
  onLayoutColsHidden?: (data: {
    currentHiddenColumnsInfo: HiddenColumnsInfo;
    hiddenColumnsDetail: HiddenColumnsInfo[];
  }) => void;
  onBeforeRender?: () => void;
  onAfterRender?: () => void;
  onMounted?: (spreadsheet: SpreadSheet) => void;
  onDestroy?: () => void;

  // ============== Resize ====================
  onLayoutResize?: (params: ResizeParams) => void;
  onLayoutResizeSeriesWidth?: (params: ResizeParams) => void;
  onLayoutResizeRowWidth?: (params: ResizeParams) => void;
  onLayoutResizeRowHeight?: (params: ResizeParams) => void;
  onLayoutResizeColWidth?: (params: ResizeParams) => void;
  onLayoutResizeColHeight?: (params: ResizeParams) => void;
  onLayoutResizeTreeWidth?: (params: ResizeParams) => void;
  onLayoutResizeMouseDown?: (data: {
    event: Partial<MouseEvent>;
    resizeInfo?: ResizeInfo;
  }) => void;
  onLayoutResizeMouseUp?: (data: {
    event: Partial<MouseEvent>;
    resizeInfo?: ResizeInfo;
  }) => void;
  onLayoutResizeMouseMove?: (data: {
    event: Partial<MouseEvent>;
    resizeInfo?: ResizeInfo;
  }) => void;

  // ============== Global ====================
  onKeyBoardDown?: (event: KeyboardEvent) => void;
  onKeyBoardUp?: (event: KeyboardEvent) => void;
  onCopied?: (copyData: string) => void;
  onActionIconHover?: (event: GEvent) => void;
  onActionIconHoverOff?: (event: GEvent) => void;
  onActionIconClick?: (event: GEvent) => void;
  onContextMenu?: (event: GEvent) => void;
  onClick?: (event: GEvent) => void;
  onHover?: (event: GEvent) => void;
  onDoubleClick?: (event: GEvent) => void;
  onMouseHover?: (event: GEvent) => void;
  onMouseUp?: (event: MouseEvent) => void;
  onMouseDown?: (event: MouseEvent) => void;
  onMouseMove?: (event: MouseEvent) => void;
  onSelected?: (cells: S2CellType[]) => void;
  onReset?: (event: KeyboardEvent) => void;
  onLinkFieldJump?: (data: { key: string; record: Data }) => void;
  onScroll?: (position: CellScrollPosition) => void;

  // ============== Auto 自动生成的 ================
  onLayoutAfterRealDataCellRender?: (options: {
    add: [number, number][];
    remove: [number, number][];
    spreadsheet: SpreadSheet;
  }) => void;
  onRowCellBrushSelection?: (event: GEvent) => void;
  onColCellBrushSelection?: (event: GEvent) => void;
}

// useResize 参数
export interface ResizeEffectParams {
  s2: SpreadSheet;
  container: HTMLDivElement; // 只包含了 sheet 容器
  wrapper: HTMLDivElement; // 包含了 sheet + foot(page) + header
  adaptive: Adaptive | undefined;
}

// Tooltip 操作项
export interface TooltipOperatorProps<Icon = Element | string, Text = string>
  extends Omit<TooltipOperatorOptions<Icon, Text>, 'onClick'> {
  onlyMenu: boolean;
  cell: S2CellType;
}

// 下钻相关类型
export interface BaseDataSet {
  name: string;
  value: string;
  type?: 'text' | 'location' | 'date';
  disabled?: boolean;
}

export interface BaseDrillDownComponentProps<DataSet = BaseDataSet> {
  className?: string;
  titleText?: string;
  searchText?: string;
  clearButtonText?: string;
  dataSet: DataSet[];
  drillFields?: string[];
  disabledFields?: string[];
  getDrillFields?: (drillFields: string[]) => void;
  setDrillFields?: (drillFields: string[]) => void;
  drillVisible?: boolean;
}

export interface PartDrillDownInfo {
  // The data of drill down
  drillData: Record<string, string | number>[];
  // The field of drill down
  drillField: string;
}

export interface PartDrillDown {
  // The configuration of drill down
  drillConfig: BaseDrillDownComponentProps;
  // The numbers of drill down result
  drillItemsNum?: number;
  fetchData: (meta: Node, drillFields: string[]) => Promise<PartDrillDownInfo>;
  // Clear the info of drill down
  clearDrillDown?: {
    rowId: string;
  };
  // Decide the drill down icon show conditions.
  displayCondition?: (meta: Node) => boolean;
}
