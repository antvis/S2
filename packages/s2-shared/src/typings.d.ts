declare module '*.svg' {
    const src: unknown;
    export default src;
}
