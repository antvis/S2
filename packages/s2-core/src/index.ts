export { Event as GEvent } from '@antv/g-canvas';
export { getTheme } from './theme';
export { Node } from './facet/layout/node';
export { Hierarchy } from './facet/layout/hierarchy';
export { buildTableHierarchy } from './facet/layout/build-table-hierarchy';
export * from './common';
export * from './utils';
export * from './cell';
export * from './sheet-type';
export * from './data-set';
export * from './interaction';
export * from './ui/tooltip';
export * from './facet/header';
