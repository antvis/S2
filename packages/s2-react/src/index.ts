// https://github.com/microsoft/TypeScript/issues/47663#issuecomment-1270716220
import type {} from 'rc-pagination';
import './utils/extendLocale';

export * from './components';
