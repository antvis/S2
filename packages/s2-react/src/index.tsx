// export * from '@antv/s2';
import React, { FC } from 'react';
import './index.less';
import './index.module.less';

export const Entry: FC = () => {
  return <div>Entry</div>;
};
