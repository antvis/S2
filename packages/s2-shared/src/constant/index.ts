export * from './i18n';
export * from './option';
export * from './resize';
export * from './sort';
export * from './classnames';
