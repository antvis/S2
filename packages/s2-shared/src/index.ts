export * from './interface';
export * from './utils/classnames';
export * from './utils/options';
export * from './utils/resize';
export * from './utils/drill-down';
export * from './utils/theme';
export * from './constant';
