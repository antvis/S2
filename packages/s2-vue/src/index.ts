import './utils/extendLocale';

export * from './hooks/useExpose';
export * from './components';
