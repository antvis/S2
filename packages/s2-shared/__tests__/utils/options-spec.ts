import { type S2Options } from '@antv/s2';
import { getBaseSheetComponentOptions } from '../../src';

describe('Options Tests', () => {
  test('should get safety options', () => {
    const options = getBaseSheetComponentOptions();

    expect(options).toMatchSnapshot();
  });

  test('should get custom options', () => {
    const tooltipOptions: S2Options = {
      tooltip: {
        enable: false,
        operation: {
          sort: false,
          menu: {
            items: [
              {
                key: 'custom',
                label: 'custom',
              },
            ],
          },
        },
      },
    };
    const options = getBaseSheetComponentOptions(tooltipOptions);

    expect(options.tooltip).toMatchInlineSnapshot(`
      Object {
        "autoAdjustBoundary": "body",
        "enable": false,
        "operation": Object {
          "hiddenColumns": true,
          "menu": Object {
            "items": Array [
              Object {
                "key": "custom",
                "label": "custom",
              },
            ],
          },
          "sort": false,
        },
      }
    `);
  });
});
