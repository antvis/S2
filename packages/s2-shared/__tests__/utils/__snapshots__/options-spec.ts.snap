// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options Tests should get safety options 1`] = `
Object {
  "conditions": Object {},
  "cornerExtraFieldText": "",
  "cornerText": "",
  "customSVGIcons": Array [],
  "debug": false,
  "device": "pc",
  "frozen": Object {
    "colCount": 0,
    "rowCount": 0,
    "rowHeader": true,
    "trailingColCount": 0,
    "trailingRowCount": 0,
  },
  "hd": true,
  "headerActionIcons": Array [],
  "height": 480,
  "hierarchyType": "grid",
  "interaction": Object {
    "autoResetSheetStyle": true,
    "brushSelection": Object {
      "colCell": true,
      "dataCell": true,
      "rowCell": true,
    },
    "copy": Object {
      "enable": true,
      "withFormat": true,
      "withHeader": false,
    },
    "eventListenerOptions": false,
    "hiddenColumnFields": Array [],
    "hoverFocus": Object {
      "duration": 800,
    },
    "hoverHighlight": true,
    "linkFields": Array [],
    "multiSelection": true,
    "overscrollBehavior": "auto",
    "rangeSelection": true,
    "resize": Object {
      "colCellHorizontal": true,
      "colCellVertical": true,
      "colResizeType": "current",
      "cornerCellHorizontal": true,
      "minCellHeight": 20,
      "minCellWidth": 20,
      "rowCellVertical": true,
      "rowResizeType": "current",
    },
    "scrollSpeedRatio": Object {
      "horizontal": 1,
      "vertical": 1,
    },
    "scrollbarPosition": "content",
    "selectedCellHighlight": false,
    "selectedCellsSpotlight": false,
  },
  "placeholder": "-",
  "seriesNumber": Object {
    "enable": false,
  },
  "showDefaultHeaderActionIcon": true,
  "style": Object {
    "colCell": Object {
      "height": 30,
      "heightByField": null,
      "maxLines": 1,
      "textOverflow": "ellipsis",
      "widthByField": null,
      "wordWrap": true,
    },
    "cornerCell": Object {
      "maxLines": 1,
      "textOverflow": "ellipsis",
      "wordWrap": true,
    },
    "dataCell": Object {
      "height": 30,
      "maxLines": 1,
      "textOverflow": "ellipsis",
      "width": 96,
      "wordWrap": true,
    },
    "layoutWidthType": "adaptive",
    "rowCell": Object {
      "heightByField": null,
      "maxLines": 1,
      "showTreeLeafNodeAlignDot": false,
      "textOverflow": "ellipsis",
      "widthByField": null,
      "wordWrap": true,
    },
    "seriesNumberCell": Object {
      "maxLines": 1,
      "textOverflow": "ellipsis",
      "wordWrap": true,
    },
  },
  "tooltip": Object {
    "autoAdjustBoundary": "body",
    "enable": true,
    "operation": Object {
      "hiddenColumns": true,
      "menu": Object {
        "items": Array [],
      },
      "sort": true,
    },
  },
  "totals": Object {},
  "width": 600,
}
`;
